packages/navigation/src/useWebScreen.tsx

```diff
@@ -0,0 +1,94 @@
+import React, { useCallback } from 'react';
+import WebScreen from './WebScreen';
+import { createNativeStackNavigator } from '@react-navigation/native-stack';
+import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
+import type { WebScreenRuleMap, WebScreenRuleConfig } from './types';
+import { isRule } from './common';
+
+type NavigatorType =
+  | ReturnType<typeof createNativeStackNavigator>
+  | ReturnType<typeof createBottomTabNavigator>;
+
+function buildWebviewComponent(
+  baseURL: string,
+  Component: React.ElementType = WebScreen
+) {
+  return (navProps: Record<string, unknown>) => (
+    <Component {...navProps} baseURL={baseURL} />
+  );
+}
+
+function findRouteInWebScreenConfig(
+  routes: WebScreenRuleMap | undefined,
+  name: string
+): WebScreenRuleMap | undefined {
+  if (!routes) {
+    return undefined;
+  }
+  for (const key of Object.keys(routes)) {
+    const route = routes[key];
+    if (key === name) {
+      if (!isRule(route)) {
+        return route?.routes;
+      }
+      return undefined;
+    }
+    if (!isRule(route)) {
+      const foundRoute = findRouteInWebScreenConfig(route?.routes, name);
+      if (foundRoute) {
+        return foundRoute;
+      }
+    }
+  }
+  return undefined;
+}
+
+function groupScreens(
+  Navigator: NavigatorType,
+  key: string | undefined,
+  config: WebScreenRuleConfig
+) {
+  const { routes } = config;
+  const { baseURL, webScreenComponent } = config;
+  const defaultComponent = buildWebviewComponent(baseURL, webScreenComponent);
+
+  const filteredRoutes = key ? findRouteInWebScreenConfig(routes, key) : routes;
+  if (filteredRoutes)
+    return (
+      <Navigator.Group>
+        {Object.entries(filteredRoutes).map(([routeName, rule]) => {
+          if (isRule(rule)) {
+            const { urlPattern, component, options } = rule;
+            return (
+              <Navigator.Screen
+                key={routeName}
+                name={routeName}
+                //  @ts-expect-error Use proper typing for main components
+                component={component ?? defaultComponent}
+                initialParams={{ baseURL, path: urlPattern }}
```

Comment:
I think we should consider how we pass in the URL to visit to the screens here. As we don't update `params` doing `replace` actions anymore, it could be that we'd be able to structure it better.

In `WebScreen` we have these lines:

```js
const path = route?.params?.path || route?.path || '';
const baseURL = route?.params?.baseURL || props?.baseURL;
```

.. which I'm not a particular fan of and would like to see refactored.

I'm wondering if it'd be possible to get the _current navigation pathname_ as a hook, similar to Expo Router's `usePathname()`?

```jsx
function WebScreen() {
  const pathname = usePathname();
  ...
}
```

Reply:
The `urlPattern` name is kinda confusing here as well as we discussed on the call. It's not _really_ a pattern (it's a special-syntax path) and it's not a url (it's a part of a path) **and** we pass it onto the screen as `params.path`